driver/pelican/types/thermSchedule.go

```diff
@@ -0,0 +1,252 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/parnurzeal/gorequest"
+	rrule "github.com/teambition/rrule-go"
+)
+
+// Login, Authentication, Thermostat ID Retrieval Structs
+type thermIDRequest struct {
+	Resources []thermIDResources `json:"resources"`
+}
+
+type thermIDResources struct {
+	Children    []thermIDChild `json:"children"`
+	GroupId     string         `json:"groupId"`
+	Permissions string         `json:"permissions"`
+}
+
+type thermIDChild struct {
+	Id          string `json:"id"`
+	Permissions string `json:"permissions"`
+}
+
+// Thermostat Settings Structs
+type settingsRequest struct {
+	Epnum    float64         `json:"epnum"`
+	Id       string          `json:"id"`
+	Nodename string          `json:"nodename"`
+	Userdata settingsWrapper `json:"userdata"`
+}
+
+type settingsWrapper struct {
+	Epnum    float64 `json:"epnum"`
+	Fan      string  `json:"fan"`
+	Nodename string  `json:"nodename"`
+	Repeat   string  `json:"repeat"`
+}
+
+// Thermostat Schedule By Day Decoding Structs
+type scheduleRequest struct {
+	ClientData scheduleSetTimes `json:"clientdata"`
+}
+
+type scheduleSetTimes struct {
+	SetTimes []scheduleTimeBlock `json:"setTimes"`
+}
+
+type scheduleTimeBlock struct {
+	HeatSetting float64 `json:"heatSetting"`
+	CoolSetting float64 `json:"coolSetting"`
+	StartValue  string  `json:"startValue"`
+	System      string  `json:"systemDisplay"`
+}
+
+// Thermostat Schedule Structs
+
+// Struct mapping each day of the week to its daily schedule
+type ThermostatSchedule struct {
+	DaySchedules map[string]ThermostatDaySchedule `msgpack:"day_schedules"`
+}
+
+// Struct containing a series of blocks that describes a one day schedule
+type ThermostatDaySchedule struct {
+	Blocks []ThermostatBlockSchedule `msgpack:"blocks"`
+}
+
+// Struct containing data defining the settings of each schedule block
+type ThermostatBlockSchedule struct {
+	CoolSetting float64 `msgpack:"cool_setting"`
+	HeatSetting float64 `msgpack:"heat_setting"`
+	System      string  `msgpack:"system"`
+	Time        string  `msgpack:"time"`
+}
+
+var week = [...]string{"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"}
+var weekRRule = [...]rrule.Weekday{rrule.SU, rrule.MO, rrule.TU, rrule.WE, rrule.TH, rrule.FR, rrule.SA}
+
+func (pel *Pelican) GetSchedule(sitename string) (map[string]ThermostatSchedule, error) {
+	// Retrieve Login Authentication Cookies
+	loginInfo := map[string]interface{}{
+		"username": pel.username,
+		"password": pel.password,
+		"sitename": sitename,
+	}
+	respLogin, _, errsLogin := gorequest.New().Post(fmt.Sprintf("https://%s.officeclimatecontrol.net/#_loginPage", sitename)).Type("form").Send(loginInfo).End()
```

Comment:
I believe we've been following the pattern of using a single `gorequest` agent for each thermostat task -- changing settings, polling the DR status, polling occupancy data, etc. -- rather than creating a fresh instance for each request.

Under this pattern, you'd add a new `gorequest` field to the `Pelican` struct and use it for all schedule polling.

If I'm understanding things right, this would also be nice because you wouldn't need to repeat the login process each time you retrieve a thermostat schedule, you just need to keep a cookie around.

Maybe you'd also be able to retain the thermostat IDs so you don't need to ask for those every time too?

Reply:
Hmm okay, yes I see that there are separate request fields for each of the tasks you mentioned. I'll update this thread with another comment when I've finished modifying the pelican.go accordingly.

Reply:
Hey @jhkolb a quick follow up here! By the most recent update, I've added a "scheduleReq" field that is the go request agent to be used for the thermostat scheduling tasks.

I had a quick question regarding the placement of the login code and retrieval of thermostatIDs. I agree with the idea that performing these requests a single time as opposed to for every request would be more efficient.

However, does that mean instead of placing the code inside the GetSchedule function, I would move it to the main.go file? Although I think login can be performed once, is there a possibility that while the schedules are being polled for the existing thermostats, a new thermostat is added, but because the thermostat IDs are only retrieved once, new thermostats are unintentionally excluded ?

Reply:
I would move the initialization code to the `NewPelican` function. That's what gets called to set up Pelican instances when the driver is first started.

How are you planning to retain the cookie and thermostat IDs? By adding fields to the Pelican struct?

Reply:
👍 sounds good. I think that would make the most sense to move it into  the NewPelican function. What do you think about the new thermostat ID issue in my previous comment?

Also, this might be a dumb question due to my lack of knowledge in security, but is there a possibility that the cookie would expire after  a certain amount of time?

Reply:
Yeah, I suppose the cookie could in theory expire, good point. So we have three options:
  1. Don't deal with cookie expiration,  let the driver exit if this occurs. (Hopefully then the XBOS infrastructure will then just restart it)
  2. Detect the error and log in again to regenerate the cookie
  3. Periodically log in to regenerate the cookie, making sure the period is shorter than the expected expiration period

As for your question about IDs, we've changed the Pelican driver so that you have to run a separate discovery tool explicitly whenever a new thermostat is added to a site. The driver itself just looks up the known thermostats and assumes this doesn't change. So you shouldn't need to worry about new thermostats here -- someone will run the discovery tool and restart the driver to deal with this.

Reply:
Per today's discussion, I implemented the following changes:
1. Added 3 new fields to the Pelican struct: Id, Cookie, and Sitename
2. Encapsulated the code for retrieving the Id, Cookie fields (which require AJAX requests) within the setCookieAndID method.
3. Added a conditional at the beginning of GetSchedule to resolve the possibility of expired cookies by calling the setCookieAndID method on monthly intervals.

These changes should resolve the ID and Cookie retrieval issue. Instead of being retrieved every time GetSchedule is called, it is now on a monthly interval. The new Pelican fields also make many of the method headers more clean and comprehensible.

If the new changes look all right, I think we can merge? :)